File: typescript/src/day2.ts
import fs from 'fs';

function getInput(): Array<string>{
    const data = fs.readFileSync('../../inputs/day2.txt', 'utf8');
    const lines = data.split('\n')

    return lines;
}

function processInput(input: Array<string>): Array<Array<number>>{
    const reports:Array<Array<number>> = [];
    for(let line of input){
        const report = line.split(' ').map(Number);
        reports.push(report)
    }

    return reports;
}

function isSafe(report:Array<number>):boolean{
    let isIncreasing = true
    let isDecreasing = true
    let isJumpOk = true
    let i = 1
    while(i < report.length && (isIncreasing || isDecreasing) && isJumpOk){
        if(report[i-1] <= report[i])
            isDecreasing = false
        if(report[i-1] >= report[i])
            isIncreasing = false
        if(Math.abs(report[i] - report[i-1]) > 3)
            isJumpOk = false
        i++
    }

    return (isIncreasing || isDecreasing) && isJumpOk
}

function solutionA(reports:Array<Array<number>>):void{
    let sum = 0
    for(let report of reports){
        if(isSafe(report))
            sum++
    }

    console.log(sum)
}

function solutionB(reports:Array<Array<number>>):void{
    let sum = 0
    for(let report of reports){
        if(isSafe(report))
            sum++
        else{
            for(let i = 0; i < report.length; i++){
                let modifiedReport = [...report]
                modifiedReport.splice(i, 1)
                if(isSafe(modifiedReport)){
                    sum++
                    break
                }
            }
        }
    }

    console.log(sum)
}


function main(){
    const input = getInput();
    const reports = processInput(input)
    solutionA(reports)
    solutionB(reports)
}

main();

File: python/day2.py
def getInput():
    data = []
    with open("../inputs/day2.txt", "r") as file:
        for line in file:
            data.append([int(x) for x in line.strip("\n").split(" ")])

    return data

def isSafe(report):
    isIncreasing = True
    isDecreasing = True
    isJumpOk = True
    i = 1
    while i < len(report) and (isIncreasing or isDecreasing) and isJumpOk:
        if report[i-1] <= report[i]:
            isDecreasing = False
        if report[i-1] >= report[i]:
            isIncreasing = False
        if abs(report[i] - report[i-1]) > 3:
            isJumpOk = False
        i += 1

    return (isIncreasing or isDecreasing) and isJumpOk


def solutionA(data):
    sum = 0
    for report in data:
        if isSafe(report):
            sum += 1

    return sum

def solutionB(data):
    sum = 0

    for report in data:
        if isSafe(report):
            sum += 1
        else:
            for i in range(len(report)):
                modifiedReport = report[:i] + report[i + 1 :]
                if isSafe(modifiedReport):
                    sum += 1
                    break

    return sum

data = getInput()
print(solutionA(data))
print(solutionB(data))